### ISSUE TITLE
Minification will remove an import that is transitively depended on

### ISSUE THREAD

**Issue by Durbatuluk1701** [JasonGross/coq-tools#211]
_CoqProject:
```
-R . Test
Test1.v
Test2.v
Test3.v
```
Test1.v:
```
Definition thing (x: nat) : nat :=
  x.
```
Test2.v:
```
Require Import Test1.
```
Test3.v:
```
Require Import Test2.

Lemma uses_thing : forall x,
  Test1.thing x = x.
Proof.
  reflexivity.
Qed.
```

Running `coq-tools/minimize-requires.py --all -f _CoqProject` will modify Test2.v by removing the `Require Import Test1.` However this will cause the project to no longer build as `Test1.thing` is no longer available in the environment.

**Comment by JasonGross** [comment]
Indeed, this is a defect that results from wanting to avoid recompiling the entire project on every change to every file.  However, we should probably add a flag `--keep-transitive-dependencies` (or similar) that runs the expensive check.  (PRs welcome!)

**Comment by Durbatuluk1701** [comment]
What would your thoughts be on a possible alternate approach where we basically have two phases of minification:
1. Directly import all transitively depended upon files
2. Work forward (i.e. in CoqProject order) with the binary search to reduce the number of imports.

If I am thinking about it correctly we would not have any need to backtrack and would still end up with a minimal number of imports.
The downside would likely be that later files would likely have a much larger search to perform, although hopefully the binary search would help with that.

**Comment by JasonGross** [comment]
Yes, this should work.  For 1, there is already code in the bug minimizer that adds all transitive `Require`s to the top of the file.  Probably we want a slightly different variant that doesn't split apart later `Require Import`s, and if we want to be extra robust, we may want to try inserting transitive `Require`s right before each existing `Require` rather than all at the top of the file.

Regarding 2, I don't think binary search helps much.  We can perhaps do binary search on the transitive `Require`s of each original `Require`ment, but in general we're searching a space of size 2<sup>n</sup> (since each `Require` can be present or not) and so cannot uniformly do much better then linear traversal of the `Require`s.  (If we have reason to believe that we'll be removing some subset more often than not, e.g., as in the case of the transitive `Require`s of each original `Require`, then we may do better by trying a couple steps of binary search locally.)

-----------

Note, however, that Step 1 is not 100% safe because global side effects of `Require` are order-dependent, so we may still want the quadratic time version.

**Comment by JasonGross** [comment]
Some incomplete notes if you (or anyone else) wants to implement this: I believe the first steps are:

1. Add a `--add-transitive-requires-before-minimizing` flag to gate this behavior
2. Add a `first_strip_requires=False` argument to disable the `strip_requires` in https://github.com/JasonGross/coq-tools/blob/d74f07b336927f4a5a5e0386c20a06f4bb713098/coq_tools/replace_imports.py#L148-L163 (note we can already pass `absolutize=()` to bypass the `get_file` normalization logic)
3. (Optional) Add a variant of https://github.com/JasonGross/coq-tools/blob/d74f07b336927f4a5a5e0386c20a06f4bb713098/coq_tools/import_util.py#L1277 that includes require locations, so we can add transitive dependencies right before each require if we want to 
4. Join requires when  `--add-transitive-requires-before-minimizing`  is passed as in https://github.com/JasonGross/coq-tools/blob/d74f07b336927f4a5a5e0386c20a06f4bb713098/coq_tools/find_bug.py#L1004-L1016
